### ISSUE TITLE
[design concern] Spellcheck for EditableText is too chatty (resource intensive)

### ISSUE THREAD

**Issue by Adam-Langley** [flutter/flutter#152205]
### Steps to reproduce

Enable spellchecking for a Text Field, and type into it quickly.
The underlying platform SpellcheckService will be called a superfluous number of times, as many of the results will be thrown away very quickly.

While supplementing the currently unimplemented spell-checker for the Windows version of our application, I discovered that while 'fetchSpellCheckSuggestions' is indeed async, it is expected to return its results quickly.
Failure to do so will cause a massive number of calls to be queued up in the async scheduler, and after you stop typing, you're waiting for them to all complete, impacting the user experience.

Try button mashing, or holding down backspace on a text field that has spell checking enabled.

To mitigate this, I implemented debouncing inside my "Win32SpellCheckService" implementation. ~I discovered that this caused rendering failures in the widget, if the user held down backspace - because when the successfully debounced call executes, the string value in the text field is now potentially shorter that the one that was spellchecked, meaning the SuggestionSpans probably reference indexes that are out of range (and thats the error reported by the EditableText widget).~

I would suggest adding debounce to the built-in implementation of EditableText, perhaps with a configuration period. It appears that all implementations of inline spell-checking I can find (Chrome, MS Word, etc) all use debouncing to avoid checking transient text that is quickly rendered dirty.

### Expected results

When typing very quickly into a text field, running of the spell checker should be debounced.

### Actual results

When typing very quickly into a text field, running of the spell checker is not debounced, causing unnecessary system load, and impacting user experience.

### Code sample

This is actually just a crude example of a win32 spell checker, with debouncing. It works fine (albeit poorly performance-wise) without debounce. Of course, it can only provide underlining, even though it returns suggestions, EditableText doesnt use them on Windows.

```dart
import 'dart:async';
import 'dart:isolate';
import 'dart:math';
import 'dart:ui';
import 'package:flutter/services.dart';
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import 'package:win32/win32.dart';

class Win32SpellCheckService implements SpellCheckService {
  final Map<String, _SpellCheckIsolate> _isolateCache = {};

  @override
  Future<List<SuggestionSpan>?> fetchSpellCheckSuggestions(Locale locale, String text) {
    final completer = Completer<List<SuggestionSpan>?>();
    final languageTag = locale.toLanguageTag();

    EasyDebounce2.debounce(
      'win32-spellcheck-debouncer',
      const Duration(milliseconds: 500),
      () {
        _processSpellCheck(languageTag, text)
          .then(completer.complete)
          .catchError(completer.completeError);
      },
      () => completer.complete([])
    );

    return completer.future;
  }

  Future<List<SuggestionSpan>?> _processSpellCheck(String languageTag, String text) async {
    if (!_isolateCache.containsKey(languageTag)) {
      await _initializeIsolate(languageTag);
    }
    
    return await _isolateCache[languageTag]!.check(text);
  }

  Future<void> _initializeIsolate(String languageTag) async {
    final receivePort = ReceivePort();
    final isolate = _SpellCheckIsolate();
    _isolateCache[languageTag] = isolate;
    
    receivePort.listen(isolate.handleResponsesFromIsolate);
    
    await Isolate.spawn(_startRemoteIsolate, receivePort.sendPort);
    final success = await isolate.initialize(languageTag);
    
    if (!success) {
      print('Spellcheck initialization failed, locale not supported: $languageTag');
    }
  }

  static void _startRemoteIsolate(SendPort port) {
    final receivePort = ReceivePort();
    port.send(receivePort.sendPort);

    Win32SpellCheckServiceWorker worker = Win32SpellCheckServiceWorker();

    receivePort.listen((dynamic message) async {
      if (message is SpellCheckIsolateMessage) {
        switch(message){
          case InitializeSpellCheckIsolateRequest(): 
            var initializeResponse = await worker.initialize(message.languageTag);
            port.send(InitializeSpellCheckIsolateResponse(initializeResponse));         
      
          case CheckSpellCheckIsolateRequest():
            var suggestions = await worker.check(message.text);
            port.send(CheckSpellCheckIsolateResponse(message.correlationId, suggestions));
        }
      }
    });
  }
}

class _SpellCheckIsolate
{
  SendPort? _sendPort;

  final Completer _isolateReadyCompleter = Completer();
  final Completer<bool> _initializeCompleter = Completer();
  final Map<String, Completer<Iterable<CheckError>>> _checkCompleters = {};
  late final String _locale;

  void handleResponsesFromIsolate(dynamic message) {
    if (message is SendPort) {
      _sendPort = message;
      _isolateReadyCompleter.complete();
    } else if (message is SpellCheckIsolateMessage) {
      switch (message){
        case InitializeSpellCheckIsolateResponse():
          if (_initializeCompleter.isCompleted != true)
          {
            _initializeCompleter.complete(message.success);
          }
        case CheckSpellCheckIsolateResponse():
          var completer = _checkCompleters[message.correlationId];
          if (completer?.isCompleted != true)
          {
            completer?.complete(message.errors);
          }
      }
    }    
  }

  Future<bool> initialize(String locale) async
  {
    _locale = locale;
    await _isolateReadyCompleter.future;    
    _sendPort!.send(InitializeSpellCheckIsolateRequest(locale));
    return await _initializeCompleter.future;
  }

  Future<List<SuggestionSpan>> check(String text) async 
  {
    await _isolateReadyCompleter.future;
    var results = <SuggestionSpan>[];
    var localeIsSupported = await _initializeCompleter.future;
    if (localeIsSupported)
    {
      Random random = Random(DateTime.now().microsecondsSinceEpoch);
      var correlation = random.nextInt(4294967295).toString();
      //var correlation = DateTime.now().microsecondsSinceEpoch.toString();
      var completer = Completer<Iterable<CheckError>>();
      _checkCompleters[correlation] = completer;
      _sendPort!.send(CheckSpellCheckIsolateRequest(correlation, text));
      var resultsMap = await completer.future;
      for(var result in resultsMap){
        var span = SuggestionSpan(
          TextRange(start: result.start, end: result.end, ),
          result.suggestions
        );
        results.add(span);
      }
    } else {
      print('Spellcheck failed, locale not supported: $_locale');
    }
    return results;
  }
}

class Win32SpellCheckServiceWorker {
  late final ISpellCheckerFactory _spellCheckerFactory;
  late final ISpellChecker _spellChecker;

  Win32SpellCheckServiceWorker() {
    CoInitializeEx(nullptr, COINIT.COINIT_MULTITHREADED);
    _spellCheckerFactory = SpellCheckerFactory.createInstance();
  }

  Future<bool> initialize(String locale) async
  {
    final languageTagPtr = locale.toNativeUtf16();
    final supportedPtr = calloc<Int32>();
    final spellCheckerPtr = calloc<COMObject>();
    
    try {
      _spellCheckerFactory.createSpellChecker(languageTagPtr, spellCheckerPtr.cast());
      _spellCheckerFactory.isSupported(languageTagPtr, supportedPtr);
      
      if (supportedPtr.value == 1) {
        _spellChecker = ISpellChecker(spellCheckerPtr);
        return true;
      } else {
        return false;  // Return empty result if language is not supported
      }
    } finally {
      free(languageTagPtr);
      free(supportedPtr);
      // We don't free spellCheckerPtr here as it's now owned by _spellCheckers
    }
  }

  Future<List<CheckError>> check(String text) async {
    var result = <CheckError>[];
    final errorsPtr = calloc<COMObject>();
    final textPtr = text.toNativeUtf16();
    try {
      final hr = _spellChecker.check(textPtr, errorsPtr.cast());
      if (!FAILED(hr)) {
        final errors = IEnumSpellingError(errorsPtr);
        await _processErrors(errors, _spellChecker, text, result);
        errors.detach();
      } else {
        print("Spellcheck native api call (ISpellChecker.check) failed with code: $hr");
      }
    } finally {
      free(textPtr);
      free(errorsPtr);
    }
    return result;
  }

  Future<void> _processErrors(IEnumSpellingError errors, ISpellChecker spellChecker, String text, List<CheckError> result) async {
    final errorPtr = calloc<COMObject>();
    int hr;
    while ((hr = errors.next(errorPtr.cast())) == S_OK) {
      final error = ISpellingError(errorPtr);
      await _processSingleError(error, spellChecker, text, result);
      error.detach();
    }
    if (FAILED(hr)){
      print("Spellcheck native api call (IEnumSpellingErrors.next) failed with code: $hr");
    }
    free(errorPtr);
  }

  Future<void> _processSingleError(ISpellingError error, ISpellChecker spellChecker, String text, List<CheckError> result) async {
    final word = text.substring(error.startIndex, error.startIndex + error.length);
   
    if (error.correctiveAction == CORRECTIVE_ACTION.CORRECTIVE_ACTION_GET_SUGGESTIONS) {
      final wordPtr = word.toNativeUtf16();
      final suggestionsPtr = calloc<COMObject>();
      try {
        final hr = spellChecker.suggest(wordPtr, suggestionsPtr.cast());
        if (!FAILED(hr)) {
          final suggestions = IEnumString(suggestionsPtr);
          final suggestionList = await _getSuggestions(suggestions);
          suggestions.detach();
          result.add(CheckError(
            error.startIndex, 
            error.startIndex + error.length,
            suggestionList,
          ));
        } else {
          print("Spellcheck native api call (ISpellChecker.getSuggestions) failed with code: $hr");
        }
      } finally {
        free(wordPtr);
        free(suggestionsPtr);
      }
    }
  }

  Future<List<String>> _getSuggestions(IEnumString suggestions) async {
    final suggestionPtr = calloc<Pointer<Utf16>>();
    final suggestionResultPtr = calloc<Uint32>();
    final suggestionList = <String>[];
    try {
      int hr;
      while ((hr = suggestions.next(1, suggestionPtr, suggestionResultPtr)) == S_OK) {
        suggestionList.add(suggestionPtr.value.toDartString());
      }
      if (FAILED(hr)){
        print("Spellcheck native api call (IEnumString.next) failed with code: $hr");
      }
    } finally {
      free(suggestionPtr);
      free(suggestionResultPtr);
    }
    return suggestionList;
  }

  void dispose() {
    _spellChecker.detach();
    _spellCheckerFactory.release();
    CoUninitialize();
  }
}


abstract class SpellCheckIsolateMessage
{
  final String correlationId;

  SpellCheckIsolateMessage(this.correlationId);
}

class InitializeSpellCheckIsolateRequest extends SpellCheckIsolateMessage
{
  final String languageTag;

  InitializeSpellCheckIsolateRequest(this.languageTag) : super('');
}

class InitializeSpellCheckIsolateResponse extends SpellCheckIsolateMessage
{
  final bool success;

  InitializeSpellCheckIsolateResponse(this.success) : super('');
}

class CheckSpellCheckIsolateRequest extends SpellCheckIsolateMessage
{
  final String text;

  CheckSpellCheckIsolateRequest(super.correlationId, this.text);
}

class CheckSpellCheckIsolateResponse extends SpellCheckIsolateMessage
{
  final List<CheckError> errors;

  CheckSpellCheckIsolateResponse(super.correlationId, this.errors);
}

class CheckError
{
  final int start;
  final int end;
  final List<String> suggestions;

  CheckError(this.start, this.end, this.suggestions);
}


/* a customised EasyDebounce which has a callback for victim invocations */

/// A void callback, i.e. (){}, so we don't need to import e.g. `dart.ui`
/// just for the VoidCallback type definition.
typedef EasyDebounceCallback = void Function();

class _EasyDebounceOperation {
  EasyDebounceCallback callback;
  EasyDebounceCallback cancelled;
  Timer timer;
  _EasyDebounceOperation(this.callback, this.cancelled, this.timer);
}

/// A static class for handling method call debouncing.
class EasyDebounce2 {
  static Map<String, _EasyDebounceOperation> _operations = {};

  /// Will delay the execution of [onExecute] with the given [duration]. If another call to
  /// debounce() with the same [tag] happens within this duration, the first call will be
  /// cancelled and the debouncer will start waiting for another [duration] before executing
  /// [onExecute].
  ///
  /// [tag] is any arbitrary String, and is used to identify this particular debounce
  /// operation in subsequent calls to [debounce()] or [cancel()].
  ///
  /// If [duration] is `Duration.zero`, [onExecute] will be executed immediately, i.e.
  /// synchronously.
  static void debounce(
      String tag, Duration duration, EasyDebounceCallback onExecute, EasyDebounceCallback onCancelled) {
    if (duration == Duration.zero) {
      _operations[tag]?.timer.cancel();
      _operations.remove(tag);
      onExecute();
    } else {
      var operation = _operations[tag];
      if (null != operation){
        operation.timer.cancel();
        operation.cancelled();
      }      

      _operations[tag] = _EasyDebounceOperation(
          onExecute,
          onCancelled,
          Timer(duration, () {
            _operations[tag]?.timer.cancel();
            _operations.remove(tag);

            onExecute();
          }));
    }
  }

  /// Fires the callback associated with [tag] immediately. This does not cancel the debounce timer,
  /// so if you want to invoke the callback and cancel the debounce timer, you must first call
  /// `fire(tag)` and then `cancel(tag)`.
  static void fire(String tag) {
    _operations[tag]?.callback();
  }

  /// Cancels any active debounce operation with the given [tag].
  static void cancel(String tag) {
    _operations[tag]?.timer.cancel();
    _operations.remove(tag);
  }

  /// Cancels all active debouncers.
  static void cancelAll() {
    for (final operation in _operations.values) {
      operation.timer.cancel();
    }
    _operations.clear();
  }

  /// Returns the number of active debouncers (debouncers that haven't yet called their
  /// [onExecute] methods).
  static int count() {
    return _operations.length;
  }
}

```

### Flutter Doctor output

[√] Flutter (Channel stable, 3.22.2, on Microsoft Windows [Version 10.0.19045.4529], locale en-NZ)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[X] Android toolchain - develop for Android devices
    X Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.10.4)
[!] Android Studio (not installed)
[√] VS Code (version 1.91.1)
[√] Connected device (3 available)
[√] Network resources

! Doctor found issues in 2 categories.

**Comment by darshankawar** [comment]
Thanks for the detailed report @Adam-Langley 
Are you seeing the reported behavior on desktop ? or on other platforms as well ?


> I would suggest adding debounce to the built-in implementation of EditableText,

There seems to be community packages offering this feature already such as https://pub.dev/packages/flutter_debouncer, https://pub.dev/packages/easy_debounce that you can check to see if it helps resolves your case or not.

**Comment by Adam-Langley** [comment]
Hi, thanks for weighing in. 

I am only currently customising/implementing the spellchecking for Windows so have only "observed" compromised performance on Windows. 

However, after digging in to the editabletext and spellcheck code files, I found a complete absence of debouncing. As such, my report is targeting the more general case - that both iOS and Android are likely to also be performing unnecessary spellchecks. This is my primary concern.  

You will see my code example already uses easydebounce to work around the lack of native debouncing - however, developers of iOS and Android apps should not have to override and "patch" the build in defaultspellcheckservice to avoid this suboptimal behaviour. 

Of course, I could invest more time to provide performance metrics on mobile, but I feel someone familiar with the could will probably know immediately if my interpretation of the code is correct. 

Perhaps if my report isn't filed in the correct way for the purposes I've just described, how should I amend it?

Thanks for your input

**Comment by darshankawar** [comment]
> I am only currently customising/implementing the spellchecking for Windows so have only "observed" compromised performance on Windows.

Spell check for EditableText on desktop is currently not supported by Flutter. See https://github.com/flutter/flutter/issues/122433

> However, after digging in to the editabletext and spellcheck code files, I found a complete absence of debouncing. As such, my report is targeting the more general case - that both iOS and Android are likely to also be performing unnecessary spellchecks. This is my primary concern.

See if using https://api.flutter.dev/flutter/material/TextField/enableSuggestions.html and setting it to `false` still gives you same behavior or not.



> but I've found this cause which is probably another bug - so I'll report it separately (#152272).

If this is for desktop only, then it is likely due to the linked issue I shared above.
